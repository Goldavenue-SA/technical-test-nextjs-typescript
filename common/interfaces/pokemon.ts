type Stats = {
  hp: number;
  attack: number;
  defense: number;
  special_attack: number;
  special_defense: number;
  speed: number;
};

interface Pokemon {
  id: number;
  name: string;
  type: string[];
  hp: number;
  attack: number;
  defense: number;
  special_attack: number;
  special_defense: number;
  speed: number;
}

type PartialPokemonProperties = Pick<Pokemon, 'id' | 'name' | 'type'>;

type PokemonTableItem = PartialPokemonProperties & {
  stats: Stats;
  power: number;
};

export type { Pokemon, PokemonTableItem };
