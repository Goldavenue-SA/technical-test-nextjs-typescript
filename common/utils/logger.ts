import winston from 'winston';

const createLogger = () => {
  if (typeof window !== 'undefined') {
    return winston.createLogger({
      level: 'info',
      format: winston.format.combine(
        winston.format.timestamp(),
        winston.format.json()
      ),
      transports: [
        new winston.transports.Console({
          format: winston.format.simple(),
        }),
      ],
    });
  }

  const logger = winston.createLogger({
    level: 'info',
    format: winston.format.combine(
      winston.format.timestamp(),
      winston.format.json()
    ),
    transports: [
      new winston.transports.Console({
        format: winston.format.simple(),
      }),
    ],
  });

  if (process.env.NODE_ENV === 'development') {
    logger.add(
      new winston.transports.File({
        filename: 'logs/error.log',
        level: 'error',
      })
    );
    logger.add(
      new winston.transports.File({
        filename: 'logs/combined.log',
      })
    );
  }

  return logger;
};

export const logger = createLogger();
