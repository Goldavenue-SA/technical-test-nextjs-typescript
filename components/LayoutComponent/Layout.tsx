import type { ReactElement } from 'react';
import { Background, Container, Main, Menu, OverlayPattern } from './Styles';
import Image from 'next/image';
import Link from 'next/link';

export const Layout = (page: ReactElement): JSX.Element => {
  return (
    <>
      <Background>
        <OverlayPattern />
        <Menu>
          <Link href={'/'} aria-label="Go to home page">
            <Image
              src="/images/pokeball.png"
              alt="Pokeball Image"
              width={96}
              height={96}
            />
            <span>Home</span>
          </Link>
        </Menu>
        <Container>
          <Main>{page}</Main>
        </Container>
      </Background>
    </>
  );
};
