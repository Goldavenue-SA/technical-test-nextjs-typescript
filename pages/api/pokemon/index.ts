import { NextApiRequest, NextApiResponse } from 'next';
import { fetchJsonFileHandler } from '../handlers/fetchJsonFileHandler';
import { errorMiddleware } from '../middleware/errorMiddleware';

export const handler = errorMiddleware(fetchJsonFileHandler);

const getPokemonData = async (req: NextApiRequest, res: NextApiResponse) => {
  return handler(req, res);
};

export default getPokemonData;
