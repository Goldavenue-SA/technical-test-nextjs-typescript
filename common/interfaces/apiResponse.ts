import { Pokemon } from './pokemon';

export interface SearchResponse {
  pokemonList: Pokemon[];
  count: number;
  min: number;
  max: number;
}
