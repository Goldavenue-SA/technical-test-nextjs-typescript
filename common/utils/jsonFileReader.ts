import path from 'path';
import fs from 'fs';
import { logger } from './logger';

export const readJsonFile = async <T>(filePath: string): Promise<T> => {
  try {
    const jsonFilePath = path.join(process.cwd(), filePath);
    const jsonData = await fs.promises.readFile(jsonFilePath, 'utf-8');

    return JSON.parse(jsonData);
  } catch (error) {
    logger.error('Error reading JSON file:', error);

    throw new Error('Failed to read or parse the JSON file.');
  }
};
