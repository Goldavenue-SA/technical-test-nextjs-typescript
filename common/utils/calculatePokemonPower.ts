import { Pokemon } from '../interfaces/pokemon';

export const calculatePokemonPower = (pokemon: Pokemon) =>
  pokemon.hp +
  pokemon.speed +
  pokemon.attack +
  pokemon.special_attack +
  pokemon.defense +
  pokemon.special_defense;
