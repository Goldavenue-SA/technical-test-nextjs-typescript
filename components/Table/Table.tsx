import { Pokemon } from '../../common/interfaces/pokemon';
import { calculatePokemonPower } from '../../common/utils/calculatePokemonPower';
import { StyledTable, TableBody, TableFooter, TableHeader } from './Styles';

export interface TableProps {
  pokemonList: Pokemon[];
}

export const Table = ({ pokemonList = [] }: TableProps) => {
  return (
    <>
      <StyledTable>
        <TableHeader>
          <tr>
            <th>ID</th>
            <th>Name</th>
            <th>Type</th>
            <th colSpan={6}>Stats</th>
            <th>Power</th>
          </tr>
        </TableHeader>
        <TableBody>
          {pokemonList.map((pokemon, index) => (
            <tr key={`${pokemon.name}${index}`}>
              <td>{pokemon.id}</td>
              <td>{pokemon.name}</td>
              <td>{pokemon.type.join(', ')}</td>
              <td>{pokemon.hp}</td>
              <td>{pokemon.speed}</td>
              <td>{pokemon.attack}</td>
              <td>{pokemon.special_attack}</td>
              <td>{pokemon.defense}</td>
              <td>{pokemon.special_defense}</td>
              <td>{calculatePokemonPower(pokemon)}</td>
            </tr>
          ))}
        </TableBody>
      </StyledTable>
    </>
  );
};
