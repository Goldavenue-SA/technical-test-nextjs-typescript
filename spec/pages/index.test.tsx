import '@testing-library/jest-dom'; // Add this import
import { render, screen } from '@testing-library/react';
import { Pokemon } from '../../common/interfaces/pokemon';
import { logger } from '../../common/utils/logger';
import { Layout } from '../../components/LayoutComponent/Layout';
import { Table } from '../../components/Table/Table';
import HomePage, { getServerSideProps } from '../../pages';

jest.mock('next/head', () => {
  return {
    __esModule: true,
    default: ({ children }: { children: React.ReactNode }) => (
      <div data-testid="mock-head">{children}</div>
    ),
  };
});

jest.mock('../../components/SearchCard/SearchCard', () => ({
  SearchCard: jest.fn(() => <div data-testid="mock-search-card" />),
}));

jest.mock('../../components/Table/Table', () => ({
  Table: jest.fn(() => <div data-testid="mock-table" />),
}));

jest.mock('../../common/utils/logger', () => ({
  logger: {
    error: jest.fn(),
  },
}));

global.fetch = jest.fn();

describe('HomePage Component', () => {
  const mockPokemonList = [
    {
      id: 1,
      name: 'Bulbasaur',
      type: ['Grass', 'Poison'],
      hp: 45,
      attack: 49,
      defense: 49,
      special_attack: 65,
      special_defense: 65,
      speed: 45,
    },
    {
      id: 2,
      name: 'Ivysaur',
      type: ['Grass', 'Poison'],
      hp: 60,
      attack: 62,
      defense: 63,
      special_attack: 80,
      special_defense: 80,
      speed: 60,
    },
  ];

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Test Component Rendering', () => {
    it('should render the HomePage with all components', () => {
      render(<HomePage pokemonList={mockPokemonList} />);

      expect(screen.getByText('Pokemon list')).toBeInTheDocument();
      expect(screen.getByTestId('mock-search-card')).toBeInTheDocument();
      expect(screen.getByTestId('mock-table')).toBeInTheDocument();
      expect(Table).toHaveBeenCalledWith(
        { pokemonList: mockPokemonList },
        expect.any(Object)
      );
    });

    it('Should render with empty pokemon list', () => {
      render(<HomePage pokemonList={[]} />);

      expect(screen.getByText('Pokemon list')).toBeInTheDocument();
      expect(Table).toHaveBeenCalledWith(
        { pokemonList: [] },
        expect.any(Object)
      );
    });

    it('Should render correct meta tags and title', () => {
      render(<HomePage pokemonList={mockPokemonList} />);

      const head = screen.getByTestId('mock-head');

      expect(head).toContainHTML(
        '<meta name="description" content="A simple Pokedex, Explore a list of Pokemons, search by name, and filter them by power threshold in this interactive Pokemon list app."'
      );

      expect(head).toContainHTML('<title>Pokemon List</title>');
      expect(head).toContainHTML('<link rel="icon" href="/favicon.ico"');
    });
  });

  describe('Test getServerSideProps', () => {
    beforeEach(() => {
      process.env.BASE_URL = 'http://localhost:3000';
    });

    it('Should fetch and return pokemon data successfully', async () => {
      (global.fetch as jest.Mock).mockResolvedValueOnce({
        json: () => Promise.resolve(mockPokemonList),
      });

      const response = await getServerSideProps({} as Pokemon[]);

      expect(global.fetch).toHaveBeenCalledWith(
        'http://localhost:3000/api/pokemon'
      );
      expect(response).toEqual({
        props: {
          pokemonList: mockPokemonList,
        },
      });
    });

    it('Should handle fetch error and return notFound', async () => {
      (global.fetch as jest.Mock).mockRejectedValueOnce(
        new Error('Failed to fetch')
      );

      const response = await getServerSideProps({} as Pokemon[]);

      expect(logger.error).toHaveBeenCalledWith(
        'Error while fetching Pokemon data'
      );
      expect(response).toEqual({ notFound: true });
    });

    it('Should handle json parsing error', async () => {
      (global.fetch as jest.Mock).mockResolvedValueOnce({
        json: () => Promise.reject(new Error('Invalid JSON')),
      });

      const response = await getServerSideProps({} as any);

      expect(logger.error).toHaveBeenCalledWith(
        'Error while fetching Pokemon data'
      );
      expect(response).toEqual({ notFound: true });
    });

    it('Should use correct base URL from environment', async () => {
      process.env.BASE_URL = 'https://custom-domain.com';

      (global.fetch as jest.Mock).mockResolvedValueOnce({
        json: () => Promise.resolve(mockPokemonList),
      });

      await getServerSideProps({} as any);

      expect(global.fetch).toHaveBeenCalledWith(
        'https://custom-domain.com/api/pokemon'
      );
    });
  });

  describe('Test Layout Integration', () => {
    it('Should have Layout assigned to getLayout', () => {
      expect(HomePage.getLayout).toBeDefined();
      expect(HomePage.getLayout).toBe(Layout);
    });
  });

  it('Should match snapshot', () => {
    const { container } = render(<HomePage pokemonList={mockPokemonList} />);
    expect(container).toMatchSnapshot();
  });
});
