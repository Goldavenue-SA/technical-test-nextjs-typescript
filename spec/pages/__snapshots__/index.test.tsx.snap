// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component Should match snapshot 1`] = `
<div>
  <div
    data-testid="mock-head"
  >
    <title>
      Pokemon List
    </title>
    <meta
      content="A simple Pokedex, Explore a list of Pokemons, search by name, and filter them by power threshold in this interactive Pokemon list app."
      name="description"
    />
    <link
      href="/favicon.ico"
      rel="icon"
    />
  </div>
  <h1>
    Pokemon list
  </h1>
  <div
    data-testid="mock-search-card"
  />
  <div
    data-testid="mock-table"
  />
</div>
`;
